Add sudo to Smartctl

Erotemic opened the pull request:
Hi, thanks for writing this tool. It saved me a lot of time.

Unfortunately I wasn't able to use it out of the box. The reason is that I don't install my Python packages in root-user space. I strictly keep everything in a local user virtual environment. This makes it tricky to execute smartctl.

Fortunately, if you simply prefix your Popen args with `sudo`, the terminal will do the "right thing" and prompt you for a sudo password if that session has not been authenticated yet. This allows the user to install pySMART without admin privileges, and request them only if needed.

The way I did this was I added a `sudo` property to the `Smartctl` class. When True is just adds sudo before calling Popen. I also made it such that if the system was non-posix, it will warn the user and ignore the setting.

I also had to make some small changes in order to make it easy for the user to enable this feature. I noticed that the `smartctl` attribute of DeviceList and Device  are defined at import time because the `SmartCtl` object is called in their `__init__` signature. This means that two global level instances are constructed anyway, so I figured why not just construct a single global instance, expose that to the user, and then use it instead. So I added a global `SMARTCTL` variable to the package, which is used as the default object for new Device and DeviceList constructions.

I added documentation for this new feature in the `__init__.py` and `README.md`. Also in the README, I changed the codeblock markdown syntax to use the one where you can specify a lexer for syntax highlighting. The "result" parts of the doctests are a bit noisy, but I think it overall looks better. Open to reverting that change though. You can preview what the new README would look like [here](https://github.com/Erotemic/py-SMART/blob/258760b75604896e45d55a49036c5c173e6d9d43/README.md). 

Lastly, I fixed a few spelling mistakes, removed some unused variables, bumped the minor version, added a CHANGELOG note.

Erotemic commented:
FYI my usage of this was to put all of the SMARTCTL tables into a pandas data frame so I could inspect all of my drives simultaneously and customize how I view the attributes:

```python
    import pandas as pd
    from pySMART import SMARTCTL
    from pySMART import DeviceList
    import ubelt as ub
    SMARTCTL.sudo = True

    devices = DeviceList()

    attr_rows = []
    dev_rows = []
    test_rows = []
    testcap_rows = []
    diag_rows = []
    msg_rows = []

    for dev in devices:
        devstate = ub.udict(dev.__getstate__(all_info=True))
        devrow = devstate - {'attributes', 'test_capabilities', 'tests', 'diagnostics', 'messages'}
        dev_rows.append(devrow)

        testcap_row = dev.test_capabilities
        testcap_row['dev'] = dev.name
        testcap_rows.append(testcap_row)

        diag_row = dev.diagnostics.__getstate__()
        diag_row['dev'] = dev.name
        diag_rows.append(diag_row)

        for test in dev.tests:
            row = {'dev': dev.name}
            row.update(test.__getstate__())
            test_rows.append(row)

        for msg in dev.messages:
            msg_rows.append({
                'dev': dev.name,
                'msg': msg,
            })

        for attr in dev.attributes:
            if attr is not None:
                row = {'dev': dev.name}
                row.update(attr.__getstate__())
                row['name'] = attr.name
                row['thresh'] = row.pop('threshold')
                attr_rows.append(row)

    dev_table = pd.DataFrame(dev_rows)
    test_df = pd.DataFrame(test_rows)
    msg_df = pd.DataFrame(msg_rows)
    testcap_df = pd.DataFrame(testcap_rows)
    attr_columns = ['dev', 'num', 'name', 'value', 'worst', 'thresh', 'type', 'updated', 'when_failed', 'raw']
    attrs_df = pd.DataFrame(attr_rows, columns=attr_columns)

    from rich import print
    print('')
    print('[green] --- Devices ---')
    print(dev_table.to_string())
    print('')
    print('[green] --- Device Test Capabilities ---')
    print(testcap_df.to_string())
    print('')
    print('[green] --- Device Tests ---')
    print(test_df.to_string())
    print('')
    print('[green] --- Device Messages ---')
    print(msg_df.to_string())
    print('')
    print('[green] --- Device Attributes ---')
    for _, group in attrs_df.groupby(['num', 'name']):
        print(group.to_string())
        value_failed = group['value'].astype(int) < group['thresh'].astype(int)
        worst_failed = group['worst'].astype(int) < group['thresh'].astype(int)
        any_failed = worst_failed | value_failed
        if any_failed.any():
            print('[red] !!! ATTENTION REQUIRED !!!')
            print(any_failed)
```

which results in something like:
![image](https://user-images.githubusercontent.com/3186211/187091893-bac602a6-3a7d-446a-817f-c737b0d3c3d0.png)

ralequi commented on README.md:
````diff
@@ -80,15 +123,19 @@ Installation
-    yum install smartmontools
+```bash
+apt-get install smartmontools
+    or
````
`#or` better if you want to be inside bash script

ralequi commented on pySMART/smartctl.py:
```diff
@@ -28,10 +28,29 @@
         self.smartctl_path = smartctl_path
         self.options: List[str] = options
+        self._sudo: bool = False
+        self.sudo = sudo
```
afaik `self.sudo = sudo` is not required if u have set a property

ralequi replied:
Ok, I suppose you want to run the setter checks here...
It's ok then...

ralequi commented on pySMART/smartctl.py:
```diff
@@ -54,10 +73,16 @@ def generic_call(self, params: List[str], pass_options=False) -> Tuple[List[str]
 
+        popen_list = []
+        if self.sudo and os.name == 'posix':
+            popen_list.append('sudo')
```
Instead of "sudo", consider usinging `['sudo','-E']` instead... just to preserve user env

ralequi replied:
From man:

```

     -E, --preserve-env
                 Indicates to the security policy that the user wishes to preserve their existing environment variables.  The security policy may return an error
                 if the user does not have permission to preserve the environment.

```

ralequi reviewed (changes requested):
I don't think the use of `sudo` is required at all inside the smartctl core.
I strongly think It is a particular case, I mean, you can keep using your local python envirorment by calling `sudo -E python <script>` or something like that.

However, as far as it seems you didn't broke anything and  provide some other minor fixes/improvements I couldn't found strong reasons for reject this PR.

Please, change those minnor request changes and change your PR to `develop` branch instead of `master` and I'll accept it.

Thanks for your contribution

Erotemic commented:
Unfortunately, `sudo -E python <script>` does not work:

```bash
(pyenv3.10.5) joncrall@toothbrush:~/misc$ sudo -E python ~/misc/notes/hardwareinfo/debug_disk.py
sudo: python: command not found
```

and it does seem that using smartctl does require superuser permissions:

```bash
(pyenv3.10.5) joncrall@toothbrush:~/misc$ smartctl /dev/sda
smartctl 7.2 2020-12-30 r5155 [x86_64-linux-5.15.0-46-generic] (local build)
Copyright (C) 2002-20, Bruce Allen, Christian Franke, www.smartmontools.org

Smartctl open device: /dev/sda failed: Permission denied
```

Perhaps there is a group a user can be added to to avoid this? That would be the best solution. I'm not a fan of using sudo in my scripts either, but in this case I don't see a way around it. I've added a note to the documentation to help users be aware of the potential issues.

I think defaulting to `-E` makes sense, but it probably also makes sense to not strictly require that. Towards that end, I've changed the `.sudo` attribute of `Smartctl` to be either `None` indicating that sudo is not requested or a `List[str]` indicating the options the user wants to pass to sudo. By default setting `self.sudo = True` will do some property magic and convert that to `['-E']` behind the scenes. I've added a doctest that demonstrates this:

```python
            >>> # xdoctest: +REQUIRES(POSIX)
            >>> self = Smartctl()
            >>> print(f'self.sudo={self.sudo}')
            >>> self.sudo = True
            >>> print(f'self.sudo={self.sudo}')
            >>> self.sudo = []
            >>> print(f'self.sudo={self.sudo}')
            >>> self.sudo = False
            >>> print(f'self.sudo={self.sudo}')
            self.sudo=None
            self.sudo=['-E']
            self.sudo=[]
            self.sudo=None
```

I've also marked it with an [xdoctest](https://github.com/Erotemic/xdoctest) directive in case you're interested in running these as part of the test suite.

ralequi commented on pySMART/smartctl.py:
```diff
@@ -28,28 +28,55 @@
 
 class Smartctl:
-    def __init__(self, smartctl_path=SMARTCTL_PATH, options: List[str] = [], sudo: bool = False):
+    def __init__(self, smartctl_path=SMARTCTL_PATH, options: List[str] = [], sudo: Union[bool | List[str]] = False):
```
I think you should change this into `Union[bool,List[str]]` in order to have compatibility with << python 3.10

see: https://docs.python.org/3.9/library/typing.html

Erotemic replied:
Good call, I can see I was thinking of that, but I used the pipe instead of a comma anyway :upside_down_face:

ralequi reviewed (changes requested):
Please, fix the python compatibility issue.

I'll merge when it's solved
